packages/server/utils/RedisLockQueue.ts

```diff
@@ -0,0 +1,80 @@
+import sleep from '../../client/utils/sleep'
+import generateUID from '../generateUID'
+import getRedis from './getRedis'
+
+export default class RedisLockQueue {
+  uid = generateUID()
+  queueKey: string
+  waitTime: number
+  ttl: number
+  queued = false
+  redis = getRedis()
+
+  constructor(key: string, ttl: number) {
+    this.queueKey = `lock:queue:${key}`
+    this.ttl = ttl
+    this.waitTime = 0
+  }
+
+  unlock = async () => {
+    // Remove all occurrences of the event from the list
+    return this.redis.lrem(this.queueKey, 0, this.uid)
+  }
+
+  checkLock = async () => {
+    // Add an event to the queue only once
+    if (!this.queued) {
+      // Use transaction to add an event to the queue and set initial LIST TTL
+      const res = await this.redis
+        .multi()
+        .rpush(this.queueKey, this.uid)
+        // TODO: Use NX option to set TTL only on LIST creation (Redis 7.0 required)
+        .pexpire(this.queueKey, this.ttl)
+        .exec()
+      this.queued = true
+      const resultingListLength = res[0][1]
+      if (resultingListLength === 1) {
+        // If the resulting LIST length is 1, we're already first on the list, return no lock
+        return false
+      }
+    }
+    // Checking if we are the first on the list
+    const head = await this.redis.lindex(this.queueKey, 0)
+    if (head == null) {
+      // If no head element exists, the queue was removed for some reason
+      // Now we cannot guarantee the correct order of the event
+      throw new Error(`Could not achieve lock for ${this.queueKey}. Queue does not exists`)
+    }
+    if (head === this.uid) {
+      // We are the first on the list, no lock
+      // Refresh TTL on the LIST before running the event
+      const newTTL = await this.redis.pexpire(this.queueKey, this.ttl)
+      if (newTTL === 0) {
+        // TTL was not set because queue disappeared
+        throw new Error(`Could not achieve lock for ${this.queueKey}. Unable to set TTL`)
+      }
+      return false
+    }
+    return true
+  }
+
+  lock = async (maxWait: number) => {
+    // Check lock every 100 ms
+    const checkAfterMs = 100
+    // "Infinity" loop
+    for (let i = 0; i < 1000; i++) {
+      const hasLock = await this.checkLock()
+      if (!hasLock) {
+        return
```

Comment:
The `ttl` means no operation should take more than 3 seconds.
In other words, if `unlock` doesn't get called in 3 seconds, it should get get called automatically

consider:

```ts
if (!hasLock) {
  this.timeoutId = setTimeout(() => this.unlock(), this.ttl)
  return
}
```

then in `unlock` we can call `clearTimeout(this.timeoutId)`

this protects us from developers accidentally not calling `unlock()`.
however, it still doesn't protect us from restarts.

what _does_ protect us from restarts is detecting when the `head` hasn't changed for > ttl :wink: 
i'll leave that as a hint & see what you come up with

Reply:
@mattkrick currently implemented solution handles both automatic `unlock` and has a protection from restarts.

* As we refresh `LIST TTL` right before starting any Event, each event would have enough time to execute.
* If process crashes or developer forgets to unlock, Redis will remove the queue by `TTL` which removes the lock and allows new events to run.
* In case of restart, events are stuck in the queue and not running, so they are not refreshing the `TTL`. Redis will remove this queue along with all stuck events.

However, it has two issues: 
1. Currently, we also refresh `TTL` on adding a new event to a queue. That means, `TTL` might unexpectedly grow. For example, if someone keep adding new events fast, it prevents the stuck queue to be removed.  (It would be possible to avoid this with Redis 7.0)
2. When unlocking the queue automatically, we remove the whole queue instead of a particular stuck events.

I'm not sure if **2** is an issue, because if we need to unlock by `TTL`, it most likely means that server restart occurred, which should be a rare case.

However, considering these issues and your comments, there are several alternative solutions to TTL:

## First solution

Add `setTimeout with unlock()` to current TTL LIST approach.

**Pros**
* This will make a single stuck event to unlock the queue after TTL without delay

**Cons**
* When invalidating a queue after restart, it may still have some delays.
* We still remove the whole queue


## Second solution

As suggested, let's detect when the head hasn't changed for > ttl.

* For this purpose, for each queue we would have a corresponding separate `KEY` with `TTL` on it.
* We refresh this `KEY` `TTL` before running the event
* If another event finds that the `KEY` is missing (was removed by TTL), it drops the whole queue.
* We do not have a `TTL` on `LIST`, as it is redundant in this case
* We do not use `setTimeout with unlock()`, as it is redundant in this case

**Pros**

* No delays when doing unlock and invalidating a queue after restart

**Cons**

* Don't we just duplicate redis functionality?
* We still remove the whole queue

## Third solution

Modify the second solution and `POP` the first event instead of removing the whole queue.

It is possible. However, there is no guarantee that other events in a queue are not stuck too (e.g. server restarted). If we `POPs` a single event, we don't know whether a node process for the next event exists. We must wait for the `TTL` again.
In this case, for example, if we have 10 events stuck in a queue, we would wait for the unlock `(10 * TTL)ms` which might be too long.

**Pros**

* This will make a single stuck event to unlock the queue after TTL without delay
* Do not removes the whole queue on unlock

**Cons**

* Invalidating a restarted queue with many events in it might take a lot of time.

## Fourth solution

Have a health check as was suggested in https://github.com/ParabolInc/parabol/issues/5268#issuecomment-904721874

> 2) Have a health check for each request

> Along with adding request to a queue, for each request we set a simple key process:UID with TTL on it.
> Each request, when checks a queue, refreshes its TTL.
> If process:UID key does not exists for the UID in a queue, we know that the corresponding process crashed.
> So each event always checks the process:UID of the first item in the queue and removes the event if necessary.

**Pros**

* Seems to solve all the previously mentioned problems

**Cons**

* Adds additional complexity

Personally, I'm not sure if **2** is a big issue, therefore keeping solution as is or adding `setTimeout with unlock()` feels to be enough.

Did I understand you correctly? What do you think?

Reply:
@mattkrick there is one issue with this approach:

There is no guarantee that `head === uidToFlush` will be checked after `100ms`. 

We have multiple tasks in a queue, and therefore multiple processes running in background. That means the check will happen at random time. It will also happen when a new event arrived.

So it is possible that head will be instantly removed before it even started.

Reply:
@mattkrick yeah, I think this might work.

Just we need to make sure to select `waitTimeMs` (currently 100ms) to be not less than time required to start a task.